### PR TITLE
ddl:check partition count limit when create  partitioned table and add partitioned table 

### PR DESCRIPTION
**Pull request by ciscoxll**
## What have you changed? (mandatory)
We limit the total partition of the table to a maximum of `1024`, so we need to add some checks.
* check create partitioned table 
* check add partitioned table

More than `1024` will cause `ErrTooManyPartitions` errors.
## What is the type of the changes? (mandatory)

New feature
- New feature (non-breaking change which adds functionality)
- Improvement (non-breaking change which is an improvement to an existing feature)
- Bug fix (non-breaking change which fixes an issue)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How has this PR been tested? (mandatory)
 Unit test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

## Does this PR need to be added to the release notes? (mandatory)
No

### CODE REVIEW

**Comment by ciscoxll** [comment]
@zimulala @winkyao @tiancaiamao @shenli PTAL.

**Comment by tiancaiamao** [comment]
LGTM

**Comment by ciscoxll** [comment]
@zimulala @winkyao PTAL.

File: ddl/partition.go [hunk 1]
```diff
@@ -216,3 +216,11 @@ func onDropTablePartition(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 	job.Args = []interface{}{partitionID}
 	return ver, nil
 }
+
+// TODO: We haven't checked subpartitions yet.
```
Comment (coocood) [inline ddl/partition.go]: This TODO will never be done, how about removing it?

Comment (tiancaiamao) [reply ddl/partition.go]: Agree.

**Comment by ciscoxll** [comment]
@coocood PTAL.

File: ddl/ddl.go [hunk 2]
```diff
@@ -172,6 +175,8 @@ var (
 	ErrTooManyValues = terror.ClassDDL.New(codeErrTooManyValues, mysql.MySQLErrName[mysql.ErrTooManyValues])
 	//ErrDropLastPartition returns cannot remove all partitions, use drop table instead.
 	ErrDropLastPartition = terror.ClassDDL.New(codeDropLastPartition, mysql.MySQLErrName[mysql.ErrDropLastPartition])
+	//ErrTooManyPartitions returns too many partitions (including subpartitions) were defined.
```
Comment (coocood) [inline ddl/ddl.go]: the `suppartition` part can be removed.

File: ddl/ddl.go [hunk 3]
```diff
@@ -61,6 +61,9 @@ var (
 	// a newly created table. It takes effect only if the Storage supports split
 	// region.
 	EnableSplitTableRegion = false
+
+	// PartitionCountLimit is limit of the number of partitions in a table.
+	PartitionCountLimit = 1024
```
[the remark this answers is not in the document]
Comment (ciscoxll) [reply ddl/ddl.go]: @zz-jason Yes.

[the remark this answers is not in the document]
Comment (zz-jason) [reply ddl/ddl.go]: OK. It would be nice if we put the link of corresponding mysql document here 😄

**Comment by coocood** [comment]
LGTM

File: ddl/ddl.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zz-jason) [reply ddl/ddl.go]: s/linking/link:/

**Pull request merged**